std/assembly/array.ts

```diff
@@ -476,6 +476,44 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    let result: valueof<T>[] = new Array<valueof<T>>(0);
+    let resultLength: i32 = 0;
+    let length = this.length;
+    let dataStart = this.dataStart;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(dataStart + (i << alignof<T>()));
+
+      if (child == null) {
+        ensureSize(changetype<usize>(result), resultLength + 1, alignof<valueof<T>>());
+        // if child == null and valueof<T> is a numeric type, it will push 0 here
+        store<valueof<T>>(result.dataStart + (<usize>resultLength << alignof<valueof<T>>()), null);
+        resultLength++;
+        continue;
+      }
+
+      let subDataStart = child.dataStart;
+      let sublength = child.length;
+      let subchild: valueof<T>;
+      for (let j = 0; j < sublength; j++) {
+        if (isManaged<valueof<T>>()) {
+          subchild = changetype<valueof<T>>(__retain(load<usize>(subDataStart + (j << alignof<valueof<T>>()))));
+        } else {
+          subchild = load<valueof<T>>(subDataStart + (j << alignof<valueof<T>>()));
+        }
+        ensureSize(changetype<usize>(result), resultLength + 1, alignof<valueof<T>>());
```

Comment:
Hmm, it seems `ensureSize` could call really often inside internal loop. I guess better precompute final array and use 2 passes:
```ts
var resultSize = 0
for (let i = 0; i < length; i++) {
  let child = unchecked(this[i]);
  if (Array.isArray(child)) {
    resultSize += child.length;
  }
}
let result: valueof<T>[] = new Array<valueof<T>>(resultSize);
for (let i = 0; i < length; i++) {
  let child = unchecked(this[i]);
  let sublength = child.length;
  for (let j = 0; j < sublength; j++) {
    let subchild = unchecked(child[j]);
    result[i] = subchild;
  }
}
```

Reply:
I'll avoid doing the unchecked gets too. Sounds good to me :)

Reply:
With the ensureSize() function removed, I opted to create the resulting array with a manual heap allocation. This enabled a lot more optimizations like using `memory.copy()` for child elements. This should be all set.